suseconnect/migration.go

```diff
@@ -0,0 +1,731 @@
+}
+
+func cleanupProductRepos(p connect.Product, force bool) error {
+	productPackages, err := connect.FindProductPackages(p.Name)
+	if err != nil {
+		return err
+	}
+	repos, err := connect.Repos()
+	if err != nil {
+		return err
+	}
+	for _, availableProduct := range productPackages {
+		// skip non-obsolete products
+		if compareEditions(availableProduct.Edition, p.Edition()) >= 0 {
+			continue
+		}
+		// filter out "(System Packages)" and already disabled repos
+		found := false
+		for _, r := range repos {
+			if r.Name == availableProduct.Repo && r.Enabled {
+				found = true
+				break
+			}
+		}
+		if !found {
+			continue
+		}
+		QuietOut.Printf("Found obsolete repository %s", availableProduct.Repo)
+		if force {
+			QuietOut.Println("... disabling.")
+			connect.DisableRepo(availableProduct.Repo)
+		} else {
+			for {
+				fmt.Printf("\nDisable obsolete repository %s [y/n] (y): ", availableProduct.Repo)
+				scanner := bufio.NewScanner(os.Stdin)
+				if !scanner.Scan() {
+					QuietOut.Print("\nStandard input seems to be closed, please use '--non-interactive' option\n")
+					os.Exit(1)
+				}
+				choice := strings.ToLower(strings.TrimSpace(scanner.Text()))
+				if interrupted {
+					return ErrInterrupted
+				}
+				if choice == "n" {
+					fmt.Print("\n")
+					break
+				} else if choice == "y" || choice == "" {
+					fmt.Println("... disabling.")
+					connect.DisableRepo(availableProduct.Repo)
+					break
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// updates system records in SCC/SMT
+// adds/removes services to match target state
+// disables obsolete repos
+// returns base product version string
+func migrateSystem(migration connect.MigrationPath, forceDisableRepos bool) (string, error) {
+	var baseProductVersion string
+
+	for _, p := range migration {
+		msg := "Upgrading product " + p.FriendlyName
+		connect.QuietOut.Println(msg)
+		service, err := connect.UpgradeProduct(p)
+		if err != nil {
+			return baseProductVersion, fmt.Errorf("%s: %v", msg, err)
+		}
+
+		if service.ObsoletedName != "" {
+			msg := "Removing service " + service.ObsoletedName
+			VerboseOut.Println(msg)
+			err = connect.MigrationRemoveService(service.ObsoletedName)
+			if err != nil {
+				return baseProductVersion, err
+			}
+		}
+
+		if err := cleanupProductRepos(p, forceDisableRepos); err != nil {
+			return baseProductVersion, err
+		}
+
+		msg = "Adding service " + service.Name
+		VerboseOut.Println(msg)
+		err = connect.MigrationAddService(service.URL, service.Name)
+		if err != nil {
+			return baseProductVersion, err
+		}
+
+		// store the base product version
+		if p.IsBase {
+			baseProductVersion = p.Version
+		}
+		if interrupted {
+			return baseProductVersion, fmt.Errorf("%s: %v", msg, ErrInterrupted)
+		}
+	}
+	return baseProductVersion, nil
+}
+
+// containsProduct returns true if given slice of products contains one with given name.
+func containsProduct(products []connect.Product, name string) bool {
+	for i := range products {
```

Comment:
I guess there is no need for an index here? Just `for _, p := range products ...`